File: api-impl-js/core/citations/quality-metrics.js
'use strict';

/**
 * Citation Quality Metrics (Citation_Application.md §13).
 *
 * ALCE-style citation recall + precision, computed over per-sentence
 * judgments produced by the citationQuality LLM pass:
 *
 *   - Citation RECALL    — of the claim-bearing sentences, how many are
 *                          actually backed by a supporting citation?
 *   - Citation PRECISION — of the citation instances in the prose, how many
 *                          genuinely support the sentence they're attached to?
 *                          (Catches a valid handle wrapped around the wrong
 *                          span — something citationApply cannot detect, since
 *                          it only checks the handle exists in the citable set.)
 *
 * The math here is PURE and deterministic; only the per-sentence judgment
 * (which sentence asserts a claim, whether its citation supports it) is the
 * non-deterministic LLM step. Splitting it this way keeps the metric
 * computation unit-testable and conformance-friendly.
 *
 * Denominators can be zero (no claim-bearing sentences, no citations). In
 * those cases the corresponding score is `null` (not 0 and not 1) so a
 * response with nothing to cite doesn't distort aggregate accuracy numbers.
 */

/**
 * @typedef {Object} SentenceJudgment
 * @property {string}  [text]              the sentence
 * @property {boolean} claim_bearing       does it assert a checkable factual claim?
 * @property {boolean} has_citation        is a <citation> attached to it?
 * @property {boolean} citation_supports   do the attached citation(s) actually support it?
 * @property {number[]} [cited_seqs]       display seqs cited on this sentence
 */

function round(n) {
  return n == null ? null : Number(n.toFixed(3));
}

/**
 * @param {SentenceJudgment[]} judgments
 * @returns {Object} quality summary
 */
function computeCitationQuality(judgments) {
  const sentences = Array.isArray(judgments) ? judgments : [];

  let claimBearing = 0;
  let claimBearingSupported = 0; // claim-bearing AND backed by a supporting citation
  let citedSentences = 0;        // sentences carrying at least one citation
  let citedSupported = 0;        // cited sentences whose citation actually supports them
  let uncitedClaims = 0;         // claim-bearing sentences with NO citation (recall gaps)
  let misattributed = 0;         // cited sentences whose citation does NOT support them (precision gaps)

  for (const s of sentences) {
    const isClaim = !!s.claim_bearing;
    const hasCite = !!s.has_citation;
    const supports = !!s.citation_supports;

    if (isClaim) {
      claimBearing++;
      if (hasCite && supports) claimBearingSupported++;
      if (!hasCite) uncitedClaims++;
    }
    if (hasCite) {
      citedSentences++;
      if (supports) citedSupported++;
      else misattributed++;
    }
  }

  const citation_recall = claimBearing > 0 ? claimBearingSupported / claimBearing : null;
  const citation_precision = citedSentences > 0 ? citedSupported / citedSentences : null;
  const f1 = (citation_recall != null && citation_precision != null && (citation_recall + citation_precision) > 0)
    ? (2 * citation_recall * citation_precision) / (citation_recall + citation_precision)
    : null;

  return {
    citation_recall: round(citation_recall),
    citation_precision: round(citation_precision),
    citation_f1: round(f1),
    counts: {
      sentences: sentences.length,
      claim_bearing: claimBearing,
      claim_bearing_supported: claimBearingSupported,
      cited_sentences: citedSentences,
      cited_supported: citedSupported,
      uncited_claims: uncitedClaims,
      misattributed_citations: misattributed,
    },
  };
}

module.exports = { computeCitationQuality };

File: api-impl-py/irg/citations/quality_metrics.py
"""Citation quality metrics (Python port of api-impl-js/core/citations/quality-metrics.js).

ALCE-style citation recall + precision over per-sentence judgments. Denominators
can be zero (no claim-bearing sentences / no citations); the corresponding score
is ``None`` (not 0, not 1) so a response with nothing to cite doesn't distort
aggregate accuracy numbers. PURE.
"""


def _round3(n):
    if n is None:
        return None
    # Match JS Number(n.toFixed(3)): format to 3 decimals then back to float.
    return float(f"{n:.3f}")


def compute_citation_quality(judgments):
    """Compute recall/precision/f1 + counts from per-sentence judgments.

    Each judgment: {claim_bearing, has_citation, citation_supports, cited_seqs?}.
    """
    sentences = judgments if isinstance(judgments, list) else []

    claim_bearing = 0
    claim_bearing_supported = 0
    cited_sentences = 0
    cited_supported = 0
    uncited_claims = 0
    misattributed = 0

    for s in sentences:
        is_claim = bool(s.get("claim_bearing"))
        has_cite = bool(s.get("has_citation"))
        supports = bool(s.get("citation_supports"))

        if is_claim:
            claim_bearing += 1
            if has_cite and supports:
                claim_bearing_supported += 1
            if not has_cite:
                uncited_claims += 1
        if has_cite:
            cited_sentences += 1
            if supports:
                cited_supported += 1
            else:
                misattributed += 1

    citation_recall = (claim_bearing_supported / claim_bearing) if claim_bearing > 0 else None
    citation_precision = (cited_supported / cited_sentences) if cited_sentences > 0 else None
    if citation_recall is not None and citation_precision is not None and (citation_recall + citation_precision) > 0:
        f1 = (2 * citation_recall * citation_precision) / (citation_recall + citation_precision)
    else:
        f1 = None

    return {
        "citation_recall": _round3(citation_recall),
        "citation_precision": _round3(citation_precision),
        "citation_f1": _round3(f1),
        "counts": {
            "sentences": len(sentences),
            "claim_bearing": claim_bearing,
            "claim_bearing_supported": claim_bearing_supported,
            "cited_sentences": cited_sentences,
            "cited_supported": cited_supported,
            "uncited_claims": uncited_claims,
            "misattributed_citations": misattributed,
        },
    }
